# Assignment 3 - Build CNN Model For Classification Of Flowers

# 1. Unzip dataset

In [9]:
!unzip '/content/Flowers-Dataset.zip'

unzip:  cannot find or open /content/Flowers-Dataset.zip, /content/Flowers-Dataset.zip.zip or /content/Flowers-Dataset.zip.ZIP.


# Importing Necessary Libraries

In [3]:
import warnings
warnings.filterwarnings("ignore")

In [4]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense,Activation,Dropout,Conv2D,Flatten,MaxPool2D,Reshape,InputLayer
from tensorflow.keras.applications.resnet50 import ResNet50
from tensorflow.keras.applications.resnet50 import preprocess_input
from tensorflow.keras.preprocessing import image
from tensorflow.keras.preprocessing.image import ImageDataGenerator,load_img,img_to_array
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau

# 2. Image Augumentation

* Dataset consist of 5 classes.
* **Daisy** - European Species of Aster family.
* **Sunflower** - Identified as the genus of Helianthus.
* **Tulip** - It belongs to the species of spring blooming geophytes.
* **Rose** - It belongs to the family of rosaceae.
* **Dandelion** - Indentifies as the genus of Asterceae.

In [5]:
path = 'flowers/'

In [6]:
train_data_gen = ImageDataGenerator(rescale = 1./255,
                             shear_range = 0.2,
                             zoom_range = 0.2,
                             horizontal_flip = True,
                             validation_split = 0.30)
test_data_gen = ImageDataGenerator(rescale = 1./255,validation_split = 0.30)

In [23]:
training_set = train_data_gen.flow_from_directory(path,
                                                 target_size=(64,64),
                                                 batch_size=100,
                                                 class_mode='categorical',
                                                 shuffle=True,
                                                 color_mode='rgb',
                                                 subset = 'training')

testing_set = test_data_gen.flow_from_directory(path,
                                                 target_size=(64,64),
                                                 batch_size=100,
                                                 class_mode='categorical',
                                                 shuffle=True,
                                                 color_mode='rgb',
                                                 subset = 'validation')

FileNotFoundError: ignored

# 3. Create the model

In [7]:
model = Sequential()

# 4. Add Layers (Convolution,MaxPooling,Flatten,Dense-Hidden Layers,Output)

In [8]:
#convolution and Pooling layer 1
model.add(Conv2D(filters=48,kernel_size=3,activation='relu',input_shape=(64,64,3)))
model.add(MaxPool2D(pool_size=2,strides=2))
model.add(Dropout(0.2))

#convolution and Pooling layer 2
model.add(Conv2D(filters=32,kernel_size=3,activation='relu'))
model.add(MaxPool2D(pool_size=2,strides=2))
model.add(Dropout(0.2))

#Flattening the images
model.add(Flatten())

#Fully Connected layers
model.add(Dense(64,activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(5,activation='softmax'))

In [11]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 62, 62, 48)        1344      
                                                                 
 max_pooling2d (MaxPooling2D  (None, 31, 31, 48)       0         
 )                                                               
                                                                 
 dropout (Dropout)           (None, 31, 31, 48)        0         
                                                                 
 conv2d_1 (Conv2D)           (None, 29, 29, 32)        13856     
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 14, 14, 32)       0         
 2D)                                                             
                                                                 
 dropout_1 (Dropout)         (None, 14, 14, 32)        0

# 5. Compiling the Model

In [12]:
model.compile(loss='categorical_crossentropy',optimizer='adam',metrics=['accuracy'])

# 6. Fitting the Model

In [13]:
early_stop = EarlyStopping(monitor='val_accuracy', 
                           patience=5,verbose=1,mode='auto')

lr = ReduceLROnPlateau(monitor='val_accuracy', 
                       factor=0.2,patience=5, 
                       min_lr=0.00001)

callback = [early_stop,lr]

Training the Model

In [14]:
result = model.fit(x=training_set, validation_data=testing_set, epochs=10)

NameError: ignored

Loss and Accuracy check using plot

In [15]:
#plot the loss
plt.plot(result.history['loss'], label='train loss')
plt.plot(result.history['val_loss'], label='val loss')
plt.legend()
plt.show()

# plot the accuracy
plt.plot(result.history['accuracy'], label='train acc')
plt.plot(result.history['val_accuracy'], label='val acc')
plt.legend()
plt.show()

NameError: ignored

# 7. Saving the Model

In [16]:
model.save('daisy.h5')

# 8. Testing the Model

In [26]:
training_set.class_indices

NameError: ignored

In [18]:
classes = ['Daisy','Dandelion','Rose','Sunflower','Tulip'] 
def testing(img):
    img = image.load_img(img,target_size=(64,64)) 
    x = image.img_to_array(img) 
    x = np.expand_dims(x,axis=0) 
    pred = np.argmax(model.predict(x)) 
    return print("Predicted class as:",classes[pred])

def img_show(img):
    img1 = image.load_img(img,target_size=(64,64)) 
    plt.imshow(img1)

In [32]:
#test1
img_show('/content/flowers/sunflower/12471443383_b71e7a7480_m.jpg')
testing('/content/flowers/sunflower/12471443383_b71e7a7480_m.jpg')

FileNotFoundError: ignored

In [28]:
#test2
img_show('/content/flowers/rose/323872063_7264e7e018_m.jpg')
testing('/content/flowers/rose/323872063_7264e7e018_m.jpg')

FileNotFoundError: ignored

In [30]:
#test3
img_show('/content/flowers/dandelion/2116997627_30fed84e53_m.jpg')
testing('/content/flowers/dandelion/2116997627_30fed84e53_m.jpg')

FileNotFoundError: ignored

In [21]:
#test4
img_show('/content/flowers/daisy/1314069875_da8dc023c6_m.jpg')
testing('/content/flowers/daisy/1314069875_da8dc023c6_m.jpg')

FileNotFoundError: ignored

In [25]:
#test5
img_show('/content/flowers/tulip/132538273_335240fe5b_n.jpg')
testing('/content/flowers/tulip/132538273_335240fe5b_n.jpg')

FileNotFoundError: ignored

# **Conclusion:**
**The dataset has about 4317 images from 5 different classes.**

1.  Each classes have more than 500 images for training the data. 
2.  30% of the data taken for validation.
3.  The accuracy of the model is around 80%.
4.  The validation accuracy is around 70%.
5.  The model is built with 2 layered convolutional   network considering 1344 trainable parameters.
6.  Testing the model with unknown images gives 95% accuracy.